File: src/1201-1300/1218 - Longest Arithmetic Subsequence of Given Difference/longest_arithmetic_subsequence_of_given_difference.go
package main

func longestSubsequence(arr []int, difference int) int {
	dp := make(map[int]int)

	for _, v := range arr {
		dp[v] = dp[v-difference] + 1
	}

	max := 0

	for _, v := range dp {
		if v > max {
			max = v
		}
	}

	return max
}

File: src/1201-1300/1218 - Longest Arithmetic Subsequence of Given Difference/longest_arithmetic_subsequence_of_given_difference.rs
impl Solution {
    pub fn longest_subsequence(arr: Vec<i32>, difference: i32) -> i32 {
        *arr.iter()
            .fold(HashMap::new(), |mut map, &i| {
                map.insert(i, map.get(&(i - difference)).unwrap_or(&0) + 1);
                map
            })
            .values()
            .max()
            .unwrap()
    }
}
